splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+
+
+class AgentProtocol(asyncio.Protocol):
+    """
+    Protocol class for interacting with a local (real) gpg agent.
+    """
+    # type hints, enable when python >= 3.6 will be everywhere...
+    # agent_responded: Optional[Future]
+    # expected_inquires: Dict[bytes, Callable]
+    # untrusted_agent_response_buffer: bytearray
+    # server: ServerProtocol
+
+    def __init__(self, server: ServerProtocol) -> None:
+        super(AgentProtocol, self).__init__()
+        self.server = server
+        self.agent_transport = None
+        # We gennerally consider the agent as trusted. But since the client can
+        # determine part of the response we handle this here as untrusted.
+        self.untrusted_agent_response_buffer = bytearray()
+        loop = asyncio.get_event_loop()
+        self.agent_responded = loop.create_future()
+        self.expected_inquires = {}
+
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        self.server.log_io('connected', b'')
+        if self.server.verbose_notifications:
+            self.server.notify('connected')
+        self.agent_transport = transport
+
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        self.server.log_io('disconnected', b'')
+        if self.server.verbose_notifications:
+            self.server.notify('disconnected')
+        self.server.client_transport.close()
+
+    def get_inquires_for_command(self, command: bytes) -> Dict[bytes, Callable]:
+        if command == b'GENKEY':
+            inquires = {
+                b'KEYPARAM': self.inquire_KEYPARAM,
+                b'PINENTRY_LAUNCHED': self.inquire_PINENTRY_LAUNCHED,
+            }
+            if self.server.allow_client_passphrase:
+                inquires[b'NEWPASSWD'] = self.inquire_NEWPASSWD
+            return inquires
+        if command == b'PKDECRYPT':
+            return {
+                b'CIPHERTEXT': self.inquire_CIPHERTEXT,
+                b'PINENTRY_LAUNCHED': self.inquire_PINENTRY_LAUNCHED,
+            }
+        if command == b'PKSIGN':
+            return {
+                b'PINENTRY_LAUNCHED': self.inquire_PINENTRY_LAUNCHED,
+            }
+        return {}
+
+    # was handle_server_response in ruby version
+    async def send_command(self, command: bytes, args: Optional[bytes]):
+        assert self.server.state == ServerState.client_command
+        self.server.state = ServerState.agent_response
+        try:
+            self.expected_inquires = self.get_inquires_for_command(command)
+            loop = asyncio.get_event_loop()
+            self.agent_responded = loop.create_future()
+            if args:
+                cmd_with_args = command + b' ' + args + b'\n'
+            else:
+                cmd_with_args = command + b'\n'
+            self.agent_write(cmd_with_args)
+            await self.agent_responded
+        finally:
+            self.server.state = ServerState.client_command
```

Comment:
Do we really wan't to change the protocol state in case of an exception?

Reply:
I wanted to keep invariant that the state after the function returns is the same as before. But since we abort connection in such case anyway, it isn't needed in practice.